Bmx.Core.Test/EventTests.cs

```diff
@@ -0,0 +1,218 @@
+		private IBmxCore<IAuthenticateState, IAuthenticatedState, IAccountState, IRoleState> _bmxCore;
+
+		private const string MockIdentityProviderName = "mockProvider";
+
+		[SetUp]
+		public void Setup() {
+			_mockIdentityProvider =
+				new Mock<IIdentityProvider<IAuthenticateState, IAuthenticatedState, IAccountState>>();
+			_mockCloudProvider = new Mock<ICloudProvider<IRoleState>>();
+
+			_bmxCore = new BmxCore<IAuthenticateState, IAuthenticatedState, IAccountState, IRoleState>(
+				_mockIdentityProvider.Object, _mockCloudProvider.Object );
+
+			_bmxCore.PromptUserName += provider => "";
+			_bmxCore.PromptUserPassword += provider => "";
+			_bmxCore.PromptMfaType += options => 0;
+			_bmxCore.PromptMfaInput += prompt => "";
+			_bmxCore.PromptAccountSelection += accounts => 0;
+			_bmxCore.PromptRoleSelection += roles => 0;
+
+			_mockIdentityProvider.Setup( m => m.Name ).Returns( MockIdentityProviderName );
+			_mockIdentityProvider.Setup( provider => provider.Authenticate( It.IsAny<string>(), It.IsAny<string>() ) )
+				.ReturnsAsync( new MockAuthenticateState() );
+			_mockIdentityProvider
+				.Setup( provider =>
+					provider.ChallengeMfa( It.IsAny<MockAuthenticateState>(), It.IsAny<int>(), It.IsAny<string>() ) )
+				.ReturnsAsync( new MockAuthenticatedState() );
+			_mockIdentityProvider
+				.Setup( provider => provider.GetAccounts( It.IsAny<MockAuthenticatedState>(), It.IsAny<string>() ) )
+				.ReturnsAsync( new MockAccountState() );
+			_mockIdentityProvider
+				.Setup( provider => provider.GetServiceProviderSaml( It.IsAny<MockAccountState>(), It.IsAny<int>() ) )
+				.ReturnsAsync( "" );
+
+			_mockCloudProvider.Setup( provider => provider.GetRoles( It.IsAny<string>() ) )
+				.Returns( new MockRoleState() );
+			_mockCloudProvider.Setup( provider => provider.GetTokens( It.IsAny<MockRoleState>(), It.IsAny<int>() ) )
+				.ReturnsAsync( new Dictionary<string, string>() );
+		}
+
+		/*
+		 * Tests if BMXCore fires input prompt events if it didn't receive values for required input at the start
+		 * (Inputs can either be at the start or interactively via prompts)
+		 */
+		[Test]
+		[Category( "PromptUsername" )]
+		public void PromptUserNameCalled() {
+			bool eventRaised = false;
+			_bmxCore.PromptUserName += providerName => {
+				eventRaised = true;
+				return "";
+			};
+			_bmxCore.Print( "org" );
+			Assert.IsTrue( eventRaised );
+		}
+
+		[Test]
+		[Category( "PromptUsername" )]
+		public void PromptUserNameProviderName() {
+			string idpName = null;
+			_bmxCore.PromptUserName += providerName => {
+				idpName = providerName;
+				return "";
+			};
+			_bmxCore.Print( "org" );
+			Assert.AreEqual( MockIdentityProviderName, idpName );
+		}
+
+		[Test]
+		[Category( "PromptPassword" )]
+		public void PromptUserPasswordCalled() {
+			bool eventRaised = false;
+			_bmxCore.PromptUserPassword += providerName => {
+				eventRaised = true;
+				return "";
+			};
+			_bmxCore.Print( "org" );
+			Assert.IsTrue( eventRaised );
```

Comment:
Is it not possible to assert what `providerName` should be? Or does it not make sense to test that?

Reply:
Doesn't make sense, as org (ex: "d2l") is a user option.